app/validation/validator.py

```diff
@@ -578,18 +579,30 @@ def _get_numeric_range_values(self, answer, answer_ranges):
         }
 
     def validate_duplicates(self, json_to_validate):
+        """
+        question_id & answer_id should be globally unique with some exceptions:
+            - within a block, ids can be duplicated across variants, but must still be unique outside of the block.
+        """
+
         special_keys = ['id']
         duplicate_errors = []
 
         for special_key in special_keys:
-            unique_items = []
-
-            for value in self._parse_values(json_to_validate, special_key):
-                if value in unique_items:
-                    duplicate_errors.append(
-                        self._error_message('Duplicate {} found. value {}'.format(special_key, value)))
-                else:
-                    unique_items.append(value)
+            items_per_block = defaultdict(set)
+
+            for path, value in self._parse_values(json_to_validate, special_key):
+                if 'blocks' in path:
+                    try:
```

Comment:
Is there a better way of handling this? Check for key existence prior to insertion?

Reply:
Potentially, I know exceptions aren't liked much around here. Looking at it, this should also be catching IndexError in case the path ends early, so the check becomes a bit more complicated.

Reply:
Actually, this only needs indexerror, and that can be handled more simply.